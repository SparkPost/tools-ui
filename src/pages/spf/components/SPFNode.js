import React from 'react';
import classNames from 'classnames';
import Icon from 'components/Icon';

import './SPFNode.scss';

function SPFNode({
  tree,
  treeId,
  domain = null,
  record,
  displayType = null,
  value = null,
  root = false,
  children = null,
  expanded = true,
  expand,
  collapse
}) {
  const label = domain ? domain : `${displayType}:${value}`;

  // TODO see if we can anchor link to panel with error
  // TODO add child type 'error'

  const panelClasses = classNames('panel', {
    'spf-tree__child': !root,
    [`spf-tree__child--${displayType}`]: displayType,
    'can-expand': children
  });

  const labelClasses = classNames('spf-tree__code', {
    [`spf-tree__code--${displayType}`]: displayType,
    'spf-tree__code--label': record // if record, use label styles
  });

  const toggleClasses = classNames({
    'fa-rotate-180': expanded
  });

  return (
    <div className={classNames('spf-tree__childWrapper', {'spf-tree': root})}>
      <div className={panelClasses} onClick={() => expanded ? collapse(treeId) : expand(treeId)}>
        <div className='panel__body'>

          <code className={labelClasses}>{ label }</code>
          {record && <code className='spf-tree__code'>{ record }</code>}
          {children && <Icon name='chevron-up' extras={toggleClasses} />}

        </div>
        <div className='spf-tree__accent' />
      </div>
      <Children tree={tree} expanded={expanded} expand={expand} collapse={collapse}>{children && children.map((childId) => tree[childId])}</Children>
    </div>
  );
}

function Children(props) {
  const { children = null, expanded } = props;

  if (!expanded || !children) {
    return null;
  }

  return (
    <div className="spf-tree__children">
      {children.map((child) => <SPFNode key={child.treeId} {...props} {...child}>{child.children}</SPFNode>)}
    </div>
  );
}

export default SPFNode;
