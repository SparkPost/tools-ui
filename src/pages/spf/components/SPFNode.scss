@import "~@sparkpost/styles/src/assets/scss/shared/config";
@import "~@sparkpost/styles/src/assets/scss/shared/functions";

.spf-tree {
  .panel {
    margin-bottom: rem(9);
    border-bottom: none;
  }

  // Typography
  &__code {
    display: block;
    background: none;
    padding: 0 rem(33) 0 0;

    color: grayscale(3);
    white-space: normal;
    word-break: break-all;

    &--label {
      font-weight: 600;
      color: grayscale(2);
    }

    // Color Coding
    &--redirect { color: color(teal);    }
    &--include  { color: color(magenta); }
  }

  &__children {
    margin-left: rem(15);
  }

  &__child, &__root {
    display: block;
    transition: 0.15s;
  }

  &__root:hover {
    cursor: pointer;
    box-shadow: shadow(2);

    & > .spf-tree__chevron {
      color: color(blue);
    }
  }

  &__child {
    position: relative;

    &.can-expand:hover {
      cursor: pointer;
      box-shadow: shadow(2);

      & > .panel__body > .spf-tree__chevron {
        color: color(blue);
      }
    }

    // These are the little lines
    &:before, &:after {
      position: absolute;
      content: '';
      background: grayscale(7);
    }

    &:before {
      top: 50%;
      right: 100%;
      left: rem(-9);
      height: 1px;
    }

    &:after {
      top: rem(-9);
      bottom: 0;
      left: rem(-9);
      width: 1px;

      // To not overflow the vertical line on last children
      .spf-tree__childWrapper:last-child > & {
        bottom: 50%;
      }
    }

    // Color Coding
    &--redirect > .spf-tree__accent { background: color(teal);    }
    &--include  > .spf-tree__accent { background: color(magenta); }
  }

  // Can't have 3 pseudo elements, and border messes up margins so...
  &__accent {
    position: absolute;
    left: 0;
    top: 0;
    bottom: 0;
    width: 3px;
  }

  &__children {
    position: relative;

    // Yo this selector is crazy
    // This is the vertical line that connects two children together if there are children in between that are expanded
    .spf-tree__childWrapper:not(:last-child) > &:after {
      position: absolute;
      content: '';
      background: grayscale(7);

      top: rem(-10);
      bottom: 0;
      left: rem(-24);
      width: 1px;
    }
  }

  &__chevron {
    position: absolute;
    right: rem(18);
    top: 50%;
    transform: translate(0, -50%) rotate(180deg);
    transition: 0.15s;

    &.is-open {
      transform: translate(0, -50%);
    }
  }
}
