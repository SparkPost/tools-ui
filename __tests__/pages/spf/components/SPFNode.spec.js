import React from 'react';
import renderer from 'react-test-renderer';
import SPFNode from 'pages/spf/components/SPFNode';

describe('SPFNode Snapshots', () => {
  let type
    , value
    , record;

  beforeEach(() => {
    type = 'mx';
    value = '_spf.erlock.homes';
    record = 'sirMXaLot';
  })

  test('baseline snapshot', () => {
    const panel = <SPFNode type={ type } value={ value } record={ record }></SPFNode>;
    expect(renderer.create(panel)).toMatchSnapshot();
  });

  test('should display domain if passed', () => {
    const panel = <SPFNode domain={'domain.com'} type={ type } value={ value } record={ record }></SPFNode>;
    expect(renderer.create(panel)).toMatchSnapshot();
  });

  test('should not show record if missing', () => {
    const panel = <SPFNode type={ type } value={ value }></SPFNode>;
    expect(renderer.create(panel)).toMatchSnapshot();
  });

});
