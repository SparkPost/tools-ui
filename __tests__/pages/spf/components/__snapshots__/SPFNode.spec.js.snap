exports[`SPFNode Snapshots baseline snapshot 1`] = `
<div
  className="spf-tree__childWrapper">
  <div
    className="panel spf-tree__child spf-tree__child--mx"
    id={undefined}
    onClick={[Function]}>
    <div
      className="panel__body">
      <span>
        <code
          className="spf-tree__code spf-tree__code--mx spf-tree__code--label">
           
          mx:_spf.erlock.homes
        </code>
      </span>
      <code
        className="spf-tree__code">
        sirMXaLot
      </code>
    </div>
    <div
      className="spf-tree__accent" />
  </div>
</div>
`;

exports[`SPFNode Snapshots should display domain if passed 1`] = `
<div
  className="spf-tree__childWrapper">
  <div
    className="panel spf-tree__child spf-tree__child--mx"
    id={undefined}
    onClick={[Function]}>
    <div
      className="panel__body">
      <span>
        <code
          className="spf-tree__code spf-tree__code--mx spf-tree__code--label">
           
          domain.com
        </code>
      </span>
      <code
        className="spf-tree__code">
        sirMXaLot
      </code>
    </div>
    <div
      className="spf-tree__accent" />
  </div>
</div>
`;

exports[`SPFNode Snapshots should display error message and id 1`] = `
<div
  className="spf-tree__childWrapper">
  <div
    className="panel spf-tree__child spf-tree__child--mx"
    id="id-123"
    onClick={[Function]}>
    <div
      className="panel__body">
      <span>
        <code
          className="spf-tree__code spf-tree__code--mx">
           
          mx:_spf.erlock.homes
        </code>
      </span>
      <p
        className="spf-tree__errorMessage">
         
        Uh oh
      </p>
    </div>
    <div
      className="spf-tree__accent" />
  </div>
</div>
`;

exports[`SPFNode Snapshots should not show record if missing 1`] = `
<div
  className="spf-tree__childWrapper">
  <div
    className="panel spf-tree__child spf-tree__child--mx"
    id={undefined}
    onClick={[Function]}>
    <div
      className="panel__body">
      <span>
        <code
          className="spf-tree__code spf-tree__code--mx">
           
          mx:_spf.erlock.homes
        </code>
      </span>
    </div>
    <div
      className="spf-tree__accent" />
  </div>
</div>
`;
